### PR TITLE
Catch exceptions on checkpoint

### PR DESCRIPTION
**Pull request by nataliab**
Calling `checkpoint` for the last processed record may fail if the lease has expired or been stolen by another worker. Let's catch the exception. 

Here's an example scenario:

1. `Worker A needed 1 leases but none were expired, so it will steal lease shardId-000000000002 from worker B`
2.  Worker B  tries to checkpoint after processing the last record and it can't `"com.amazonaws.services.kinesis.clientlibrary.exceptions.ShutdownException: Can't update checkpoint - instance doesn't hold the lease for this shard"`
3. Checkpoint throwing exception causes ` [RecordProcessor-0000] c.a.s.k.c.lib.worker.ProcessTask - ShardId shardId-000000000002: Skipping over the following data records: [...]`

Example from `marketing-gateway-actors`:
https://service.us2.sumologic.com/ui/index.html#/search/54d39f62_b04e_9053_af49_bb60c00026b8

### CODE REVIEW

**Comment by jlogeart** [comment]
What is the consequence for the client in terms of consumption of events?

**Comment by jlogeart** [comment]
To expand on the previous open comment.

Currently, if an exception is thrown by `processRecords`, the events are skipped by the client.

After the change, the newly caught (and swallowed) exception will not trigger a "skip" and the events will be retried over and over again.

This is not a bad idea, but definitely a change from the existing that may lead some user of the library to stall, force the issue to be addressed manually while it currently discards the events and keeps going.

How about instead, adding a retry logic similar to https://github.com/flowcommerce/lib-event/pull/452 ? And maybe an argument to control the behavior in case of exceptions?

Note that there two different exceptions: 
 - in the `f` function
 - in the checkpoint (network error, lost lease)

They may need to be addressed differently.

**Comment by nataliab** [comment]
Here I am only focusing on `checkpointing` and the case of lost lease.

It's possible, and normal, that a checkpoint will fail if the lease is stolen for any reason.
(Error message: "Can't update checkpoint - instance doesn't hold the lease for this shard")

In the case of a lost lease the exception is thrown after the events have been processed by our `f` function (we call checkpoint for the last processed record) so I don't think that the message from KCL `Skipping over the following data records: [...]` is relevant. We haven't checkpointed, so DynamoDB is not updated. 
If the lease renewer finds that the record processor doesn't hold the lease anymore the record processor will be terminated (calling shutdown; checkpointing will fail here again).

The next worker to get the lease for this shard will start over from the last recorded checkpoint which may cause some events to be processed twice. But we have no way to prevent this as the lease is lost already. 

I'm not even sure if it's worth logging the inability to checkpoint in case of lost lease as an error or warning given that it is a normal situation. 

Retry logic from #452 makes a lot of sense for exceptions in the `f` function - different concern from checkpointing.

**Comment by jlogeart** [comment]
I agree with you :)

What are we trying to solve? 

Looking at the client implementation, we are simply moving the log from AWS SDK to this lib. (But that is only after looking after how all exceptions are handled by the sdk)

**Comment by nataliab** [comment]
I pushed an updated version following the guidelines from an example processor implementation: 
https://github.com/aws/aws-sdk-java/blob/master/src/samples/AmazonKinesis/AmazonKinesisApplicationSampleRecordProcessor.java

Also, we should only call checkpoint on shutdown for `ShutdownReason.TERMINATE`.

Benefits:

1. Logging checkpoint issues at info/warning level where appropriate.
2. No more misleading `Skipping over the following data records: [...]` messages in case of common/normal checkpoint issues. 
3. No more checkpoint exceptions for `ZOMBIE` state (lost lease).

Overall cleaner logs and easier troubleshooting/monitoring.

File: app/io/flow/event/v2/KinesisConsumer.scala [hunk 1]
```diff
@@ -120,20 +120,44 @@ case class KinesisRecordProcessor[T](
...
   override def shutdown(input: ShutdownInput): Unit = {
     logger_.withKeyValue("reason", input.getShutdownReason.toString).info("Shutting down")
+    if (input.getShutdownReason == ShutdownReason.TERMINATE) {
+      handleCheckpoint(() => input.getCheckpointer.checkpoint())
+    }
+  }
+
+  private def handleCheckpoint(checkpoint: () => Unit, retries: Int = 0): Unit = {
+
+    val MaxRetries = 8
```
Comment (jlogeart) [inline app/io/flow/event/v2/KinesisConsumer.scala]: Let's move those to the companion object

File: app/io/flow/event/v2/KinesisConsumer.scala [hunk 2]
```diff
@@ -120,20 +120,44 @@ case class KinesisRecordProcessor[T](
...
+    val BackoffTimeInMillis = 3000L
```
Comment (jlogeart) [inline app/io/flow/event/v2/KinesisConsumer.scala]: Is there a reason why 3s?

File: app/io/flow/event/v2/KinesisConsumer.scala [hunk 3]
```diff
@@ -120,20 +120,44 @@ case class KinesisRecordProcessor[T](
 
     kinesisRecords.lastOption.foreach { record =>
       logger_.withKeyValue("checkpoint", record.getSequenceNumber).info("Checkpoint")
-      input.getCheckpointer.checkpoint(record)
+      handleCheckpoint(() => input.getCheckpointer.checkpoint(record))
```
Comment (jlogeart) [inline app/io/flow/event/v2/KinesisConsumer.scala]: Isn't `input.getCheckpointer.checkpoint()` equivalent to checkpoint the last one?

If so, I suggest to simplify the `handleCheckpoint` function by checkpointing one way only.

File: app/io/flow/event/v2/KinesisConsumer.scala (hunk 2 above)
Comment (nataliab) [reply app/io/flow/event/v2/KinesisConsumer.scala]: Copied from the sample implementation:
https://github.com/aws/aws-sdk-java/blob/master/src/samples/AmazonKinesis/AmazonKinesisApplicationSampleRecordProcessor.java#L41

File: app/io/flow/event/v2/KinesisConsumer.scala (hunk 1 above)
Comment (nataliab) [reply app/io/flow/event/v2/KinesisConsumer.scala]: Yup, I can move this to 

```
object KinesisRecordProcessor {
  // Yes, it is arbitrary
  private val MaxRetries = 8
}
```

once your PR is merged

Comment (jlogeart) [reply app/io/flow/event/v2/KinesisConsumer.scala]: Feel free to move it before other PR is merged. I am not sure yet about the other PR :)

File: app/io/flow/event/v2/KinesisConsumer.scala (hunk 3 above)
Comment (nataliab) [reply app/io/flow/event/v2/KinesisConsumer.scala]: Good point! Updated!

File: app/io/flow/event/v2/KinesisConsumer.scala (hunk 1 above)
Comment (nataliab) [reply app/io/flow/event/v2/KinesisConsumer.scala]: Implemented in the latest update.

**Pull request merged**